src/pooling.jl

```diff
@@ -0,0 +1,123 @@
+    end
+end
+
+
+# Our strategy for pooling is to reshape to an array with three spatial dimensions, which
+# makes things MUCH EASIER for us on the backend side, and is in general pretty fast,
+# since we can specialize on sizes.
+for front_name in (:maxpool, :meanpool)
+    for backend in (Symbol(), :_direct)
+        for N in (3, 4)
+            @eval begin
+                function $(Symbol("$(front_name)$(backend)!"))(
+                        y::AbstractArray{T,$N}, x::AbstractArray{T,$N},
+                        pdims::PoolDims; kwargs...) where {T}
+                    $(Symbol("$(front_name)$(backend)!"))(
+                        insert_singleton_spatial_dimension(y, $(5 - N)),
+                        insert_singleton_spatial_dimension(x, $(5 - N)),
+                        insert_singleton_spatial_dimension(pdims, $(5 - N));
+                        kwargs...
+                    )
+
+                    # We explicitly return `y` here, because the backend call
+                    # itself may return a reshaped view, which we don't want.
+                    return y
+                end
+
+                # backprops too
+                function $(Symbol("∇$(front_name)$(backend)!"))(
+                        dx::AbstractArray{T,$N}, dy::AbstractArray{T,$N},
+                        x::AbstractArray{T,$N}, pdims::PoolDims;
+                        kwargs...) where {T}
+                    $(Symbol("∇$(front_name)$(backend)!"))(
+                        insert_singleton_spatial_dimension(dx, $(5 - N)),
+                        insert_singleton_spatial_dimension(dy, $(5 - N)),
+                        insert_singleton_spatial_dimension(x, $(5 - N)),
+                        insert_singleton_spatial_dimension(pdims, $(5 - N));
+                        kwargs...
+                    )
+
+                    # We explicitly return `dx` here, because the backend call
+                    # itself may return a reshaped view, which we don't want.
+                    return dx
+                end
+            end
+        end
+    end
+end
+
+
+# Finally, let's generate auto-allocating versions of all our functions, for all backends:
+for backend in (Symbol(), :_direct, :_im2col)
+    # First make auto-allocating versions of the basic pooling calls:
+    for name in (:maxpool, :meanpool)
+        @eval begin
+            function $(Symbol("$(name)$(backend)"))(
+                    x::AbstractArray{xT,N}, pdims::PoolDims; kwargs...) where {xT, N}
+                y = zeros(xT,  output_size(pdims)..., channels_out(pdims), size(x, N))
+                return $(Symbol("$(name)$(backend)!"))(y, x, pdims; kwargs...)
+            end
+            
+            # Backprops too
+            function $(Symbol("∇$(name)$(backend)"))(
+                    dy::AbstractArray{T,N}, x::AbstractArray{T},
+                    pdims::PoolDims; kwargs...) where {T, N}
+                dx = zeros(T, input_size(pdims)..., channels_out(pdims), size(dy, N))
```

Comment:
```suggestion
                dx = zeros(T, input_size(pdims)..., channels_in(pdims), size(dy, N))
```
Doesn't make a difference as `in_channels = out_channels` for pooling but dx should have dimensions of input x.

Reply:
You're right; I like this.  Thanks!